Marlin/src/gcode/calibrate/G425.cpp

```diff
@@ -0,0 +1,690 @@
+ *   UNCERTAIN - Measurement may be uncertain due to backlash
+ *   CERTAIN   - Measurement obtained with backlash compensation
+ */
+
+#ifndef CALIBRATION_MEASUREMENT_UNKNOWN
+  #define CALIBRATION_MEASUREMENT_UNKNOWN   5.0 // mm
+#endif
+#ifndef CALIBRATION_MEASUREMENT_UNCERTAIN
+  #define CALIBRATION_MEASUREMENT_UNCERTAIN 1.0 // mm
+#endif
+#ifndef CALIBRATION_MEASUREMENT_CERTAIN
+  #define CALIBRATION_MEASUREMENT_CERTAIN   0.5 // mm
+#endif
+
+#define HAS_X_CENTER (ENABLED(CALIBRATION_CUBE_MEASURE_LEFT)  && ENABLED(CALIBRATION_CUBE_MEASURE_RIGHT))
+#define HAS_Y_CENTER (ENABLED(CALIBRATION_CUBE_MEASURE_FRONT) && ENABLED(CALIBRATION_CUBE_MEASURE_BACK))
+
+#if ENABLED(BACKLASH_GCODE)
+  extern float backlash_distance_mm[], backlash_correction, backlash_smoothing_mm;
+#endif
+
+enum side_t : uint8_t { TOP, RIGHT, FRONT, LEFT, BACK, NUM_SIDES };
+
+struct measurements_t {
+  static const float dimensions[XYZ];
+  static constexpr float true_center[XYZ] = CALIBRATION_CUBE_CENTER;
+  float cube_center[XYZ] = CALIBRATION_CUBE_CENTER;
+  float cube_side[NUM_SIDES];
+
+  float backlash[NUM_SIDES];
+  float pos_error[XYZ];
+
+  float nozzle_outer_dimension[2] = {CALIBRATION_NOZZLE_OUTER_DIAMETER, CALIBRATION_NOZZLE_OUTER_DIAMETER};
+};
+
+const float measurements_t::dimensions[]  = CALIBRATION_CUBE_DIMENSIONS;
+
+/**
+ * A class to save and change the endstop state,
+ * then restore it when it goes out of scope.
+ */
+class TemporaryEndstopsState {
+  private:
+    bool saved_endstops_global_enabled, saved_soft_endstops_enabled;
+
+  public:
+    TemporaryEndstopsState(bool enable) {
+      saved_endstops_global_enabled = endstops.global_enabled();
+      endstops.enable_globally(enable);
+
+      saved_soft_endstops_enabled = soft_endstops_enabled;
+      soft_endstops_enabled = enable;
+    }
+
+    ~TemporaryEndstopsState() {
+      soft_endstops_enabled = saved_soft_endstops_enabled;
+      endstops.enable_globally(saved_endstops_global_enabled);
+    }
+};
+
+/**
+ * A class to save and change the backlash compensation state,
+ * then restore it when it goes out of scope.
+ */
+class TemporaryBacklashCompensation {
+  #if ENABLED(BACKLASH_GCODE)
+    float saved_backlash_correction;
+    #ifdef BACKLASH_SMOOTHING_MM
+      float saved_backlash_smoothing_mm;
+    #endif
+  #endif
+
+  public:
+    TemporaryBacklashCompensation(const bool enable) {
+      #if ENABLED(BACKLASH_GCODE)
+        saved_backlash_correction = backlash_correction;
+        #ifdef BACKLASH_SMOOTHING_MM
+          saved_backlash_smoothing_mm = backlash_smoothing_mm;
+          backlash_smoothing_mm = 0;
+        #endif
+        backlash_correction = enable;
+      #endif
+    }
+
+    ~TemporaryBacklashCompensation() {
+      #if ENABLED(BACKLASH_GCODE)
+        backlash_correction = saved_backlash_correction;
+        #ifdef BACKLASH_SMOOTHING_MM
+          backlash_smoothing_mm = saved_backlash_smoothing_mm;
+        #endif
+      #endif
+    }
+};
```

Comment:
I like this approach to saving values that will be restored automagically when the instance goes out of scope. We should use this technique in more places. I believe it can be generalized, so you create an instance with simply a reference to the variable being saved. The current value is stored, the new (optional) value is applied, and when it goes out of scope, it restores the saved value to the referenced variable.

I'll build this as a utility template class, play around with it, and see how it performs.

Reply:
@thinkyhead: Yes, this is a technique I learned in a C++ book, I forget which. It's particularly useful for code which makes use exceptions, where a function could get interrupted before it would have a chance to clean up after itself. Of course, Marlin does not use exceptions, but the trick can still be useful to temporarily toggles features on and off without having to remember to restore them.

Reply:
@thinkyhead: You modified the TemporaryBacklashCompensation class like this:

```
class TemporaryBacklashCompensation {
  public:
    TemporaryBacklashCompensation(const bool enable) {
      #if ENABLED(BACKLASH_GCODE)
        REMEMBER(backlash_correction);
        #ifdef BACKLASH_SMOOTHING_MM
          REMEMBER(backlash_smoothing_mm);
          backlash_smoothing_mm = 0;
        #endif
        backlash_correction = enable;
      #endif
    }
};
```

This is incorrect because the scope of the template that REMEMBER() instantiates is now constructor itself, so the value gets restored immediately when the constructor ends, rather than on the destructor as before. The correct way to compose the REMEMBER() macros would be like this:

```
class TemporaryBacklashCompensation {
  #if ENABLED(BACKLASH_GCODE)
  REMEMBER(backlash_correction);
    #ifdef BACKLASH_SMOOTHING_MM
      REMEMBER(backlash_smoothing_mm);
    #endif
  #endif
  
  public:
    TemporaryBacklashCompensation(const bool enable) {
      #if ENABLED(BACKLASH_GCODE)
        #ifdef BACKLASH_SMOOTHING_MM
          backlash_smoothing_mm = 0;
        #endif
        backlash_correction = enable;
      #endif
    }
};
```

Which is about as gnarly as it was before, so nothing much gained...

Reply:
Also, this macro:

```
#define REMEMBER(X) restorer<typeof(X)> X##_restorer(X)
```

Can be written as:

```
#define REMEMBER(X) restorer X##_restorer(X)
```

Since the compiler can infer the type of the template automatically.

Reply:
Actually, still working on this. I thought I was compiling the code, but I wasn't. The above two comments are invalid.